dace/sdfg/construction_utils.py

```diff
@@ -0,0 +1,764 @@
+            copydesc = copy.deepcopy(datadesc)
+            copydesc.transient = False
+            parent_sdfg.add_datadesc(name=data_access, datadesc=copydesc)
+
+        # Collect enclosing map scopes to route data through
+        parent_scopes = []
+        cur_parent_scope = nsdfg_node
+        scope_dict = parent_graph.scope_dict()
+        while scope_dict[cur_parent_scope] is not None:
+            parent_scopes.append(scope_dict[cur_parent_scope])
+            cur_parent_scope = scope_dict[cur_parent_scope]
+
+        # Helper: choose between full or exact-subset memlet
+        def _get_memlet(it_id: int, data_access: str, datadesc: dace.data.Data):
+            if add_with_exact_subset:
+                return dace.memlet.Memlet(data=data_access, subset=copy.deepcopy(exact_subset))
+            else:
+                return dace.memlet.Memlet.from_array(data_access, datadesc)
+
+        # --- Add input connection path ---
+
+        state = {
+            'cur_in_conn_name': f"IN_{data_access}_p",
+            'cur_out_conn_name': f"OUT_{data_access}_p",
+            'cur_name_set': False,
+        }
+
+        def _get_in_conn_name(dst, state=state):
+            if state['cur_name_set'] is False:
+                i = 0
+                while (state['cur_in_conn_name'] in dst.in_connectors
+                       or state['cur_out_conn_name'] in dst.out_connectors):
+                    state['cur_in_conn_name'] = f"IN_{data_access}_p_{i}"
+                    state['cur_out_conn_name'] = f"OUT_{data_access}_p_{i}"
+                    i += 1
+                state['cur_name_set'] = True
+
+            inner_data_access = data_access if nsdfg_connector_name is None else nsdfg_connector_name
+
+            if isinstance(dst, dace.nodes.AccessNode):
+                return None
+            elif isinstance(dst, dace.nodes.NestedSDFG):
+                return inner_data_access
+            else:
+                return state['cur_in_conn_name']
+
+        def _get_out_conn_name(src, state=state):
+            if state['cur_name_set'] is False:
+                i = 0
+                while (state['cur_in_conn_name'] in src.in_connectors
+                       or state['cur_out_conn_name'] in src.out_connectors):
+                    state['cur_in_conn_name'] = f"IN_{data_access}_p_{i}"
+                    state['cur_out_conn_name'] = f"OUT_{data_access}_p_{i}"
+                    i += 1
+                state['cur_name_set'] = True
+
+            inner_data_access = data_access if nsdfg_connector_name is None else nsdfg_connector_name
+            if isinstance(src, dace.nodes.AccessNode):
+                return None
+            elif isinstance(src, dace.nodes.NestedSDFG):
+                return inner_data_access
+            else:
+                return state['cur_out_conn_name']
+
+        an = parent_graph.add_access(data_access)
+        src = an
+        for it_id, parent_scope in enumerate(reversed(parent_scopes)):
```

Comment:
I didn't follow the logic here carefully, but couldn't we simplify the method by employing `add_memlet_path`? I see that there is also logic concerning the connectors, but I believe we also have utility methods/passes that fix them.